File: client-java/postgres-test-utils/src/main/java/org/evomaster/client/java/postgres/test/utils/PostgresContainerUtils.java
package org.evomaster.client.java.postgres.test.utils;

import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.wait.strategy.Wait;

import java.util.Collections;

/**
 * Shared utility for creating and configuring PostgreSQL Testcontainers instances.
 * Used across modules (client-java/sql, core, e2e tests) to avoid code duplication.
 */
public class PostgresContainerUtils {

    /**
     * Default version of PostgreSQL used when none is explicitly specified.
     */
    public static final String DEFAULT_POSTGRES_VERSION = "14";

    /**
     * Creates a new PostgreSQL container using the default version ({@value DEFAULT_POSTGRES_VERSION}).
     *
     * @return A configured {@link GenericContainer} instance ready to be started.
     */
    public static GenericContainer<?> newContainer() {
        return newContainer(DEFAULT_POSTGRES_VERSION);
    }

    /**
     * Creates a new PostgreSQL container using the specified version.
     * The container is pre-configured with:
     * <ul>
     *   <li>Port 5432 exposed</li>
     *   <li>Temporary filesystem for the data directory (avoids disk I/O in tests)</li>
     *   <li>{@code POSTGRES_HOST_AUTH_METHOD=trust} (no password required)</li>
     *   <li>Readiness wait on the "database system is ready to accept connections" log message</li>
     * </ul>
     *
     * @param version The PostgreSQL Docker image version tag (e.g. {@code "14"}, {@code "17"}).
     * @return A configured {@link GenericContainer} instance ready to be started.
     */
    public static GenericContainer<?> newContainer(String version) {
        return new GenericContainer<>("postgres:" + version)
                .withExposedPorts(5432)
                .withTmpFs(Collections.singletonMap("/var/lib/postgresql/data", "rw"))
                .withEnv("POSTGRES_HOST_AUTH_METHOD", "trust")
                /*
                 * A call to getConnection() when the postgres container is still not ready
                 * signals a PSQLException with message "FATAL: the database system is starting up".
                 * See: https://github.com/testcontainers/testcontainers-java/issues/317
                 */
                .waitingFor(Wait.forLogMessage(".*database system is ready to accept connections.*\\s", 2));
    }

    /**
     * Constructs a JDBC URL for a PostgreSQL database hosted in the given container.
     *
     * @param container A running PostgreSQL {@link GenericContainer} with port 5432 mapped.
     * @return JDBC URL of the form {@code jdbc:postgresql://<host>:<port>/postgres}.
     */
    public static String getJdbcUrl(GenericContainer<?> container) {
        return "jdbc:postgresql://" + container.getHost() + ":" + container.getMappedPort(5432) + "/postgres";
    }
}

File: core/src/test/kotlin/org/evomaster/core/PostgresContainerUtils.kt
package org.evomaster.core

import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy
import java.util.*

object PostgresContainerUtils {

    /**
     * Default version of PostgreSQL used for testing and database-related operations.
     * This constant is utilized in scenarios where no specific PostgreSQL version is explicitly provided,
     * ensuring a consistent version is used across various tests and utilities.
     */
    private const val DEFAULT_POSTGRES_VERSION = "14"

    /**
     * Creates and configures a new PostgreSQL container.
     *
     * @param version Specifies the PostgreSQL version to be used for the container. Defaults to `DEFAULT_POSTGRES_VERSION`.
     * @return A configured instance of `KGenericContainer` running the specified PostgreSQL version.
     */
    fun newContainer(version: String = DEFAULT_POSTGRES_VERSION): KGenericContainer {
        return KGenericContainer("postgres:$version")
            .withExposedPorts(5432)
            //https://www.postgresql.org/docs/current/auth-trust.html
            .withEnv("POSTGRES_HOST_AUTH_METHOD", "trust")
            .withTmpFs(Collections.singletonMap("/var/lib/postgresql/data", "rw"))
            /*
             * A call to getConnection() when the postgres container is still not ready
             * signals a PSQLException with message "FATAL: the database system is starting up".
             * See: https://github.com/testcontainers/testcontainers-java/issues/317
             */
            .waitingFor(
                LogMessageWaitStrategy()
                    .withRegEx(".*database system is ready to accept connections.*\\s")
                    .withTimes(2)
            )
    }

    /**
     * Constructs a JDBC URL for a PostgreSQL database based on the provided container's host and mapped port.
     *
     * @param container the container from which the host and mapped port will be obtained
     * @return a JDBC URL string for connecting to the PostgreSQL database
     */
    fun getJdbcUrl(container: KGenericContainer): String {
        val host = container.host
        val port = container.getMappedPort(5432)!!
        return "jdbc:postgresql://$host:$port/postgres"
    }
}
